Allow alternatives/any.when functionality

rsmylskibc opened the pull request.

rsmylskibc committed:
Add usage of when

- Added the ability to use/parse when conditionals
- Added tests for when usage

rsmylskibc committed:
Merge branch 'master' into rs/allow-alternatives-when-functionality

rsmylskibc commented:
Adding this functionality - https://github.com/hapijs/joi/blob/v13.1.0/API.md#anywhencondition-options
Which will allow for adding conditional requirement to validation logic

rsmylskibc commented on test/lib/whenParser.test.js:
```diff
@@ -0,0 +1,128 @@
+
+            const [ condition, options ] = whenParser.parse( config, engine );
+
+            expect( engine.string.calledOnce ).to.be.true;
```
The test verification doesn't seem to be the greatest way to prove it's working, or returns what you would expect it to. I'm still trying to think up a better way to show that the shape of options is what is expected.

rsmylskibc committed:
Update whenParser.test.js

- Added verification of condition
- Added verification of options shape

gburtini commented:
I just wanted to call out the new tests here as really well done. 

I have something I don't fully understand about this that I'll message you in Slack.

gburtini commented on lib/base.js:
```diff
@@ -14,19 +15,33 @@ class BaseSchema {
+
+                state.schema = state.schema[ key ]( condition, options );
+            }
+            else if( value === null ) {
```
Since this was always sort of unclear, I'd recommend adding a comment here indicating what `state.schema[key]` is and why it is parsed uniquely in the `when` case.

gburtini commented on lib/base.js:
```diff
@@ -14,19 +15,33 @@ class BaseSchema {
+            if( key === 'when' ) {
 
-            if( value === null ) {
+                const [ condition, options] = whenParser.parse(value, engine);
```
Nit: There's a missing space on the right side of this expansion.

rsmylskibc commented on lib/whenParser.js:
```diff
@@ -0,0 +1,158 @@
+            }
+        }
+        else if (utils.isObject(options)) {
+            let schema = engine;
```
I wanted to clone this, but I didn't find a good way to deep clone. Since `engine` is essentially `Joi`, calls to Joi clone the object, it should be separating itself from the original as soon as the first call is made.

gburtini replied:
`cloneDeep`: https://www.npmjs.com/package/clone-deep

Also available in Lodash.

rsmylskibc replied:
Do you think it's a good idea to add clone-deep? Depending on the shape of the condition, we could clone a few times, in each full whenParser.parse(), but most likely it will only happen once or twice

gburtini replied:
I think it is unnecessary here.

rsmylskibc committed:
Update formatting and add more comments

rsmylskibc committed:
Fix missed formatting updates

rsmylskibc committed:
Rename lib to src

rsmylskibc committed:
Merge branch 'master' into rs/allow-alternatives-when-functionality

# Conflicts:
#	package-lock.json
#	src/base.js
#	src/utils.js

rsmylskibc committed:
Merge branch 'master' into rs/allow-alternatives-when-functionality

rsmylskibc committed:
Add clone-deep package

- Added clone-deep package
- Used clone-deep to fully clone engine when applying when config options
